### DISCUSSION TITLE
graphql-java imports sun.misc in the MANIFEST.MF of v 17.4

### DISCUSSION THREAD

**Discussion by schaefa** [graphql-java/graphql-java#3089] [General]
Hi

I ran into a problem when using graphql-java with Sling's Feature Models. Building the feature models it complains about sun.misc being imported by graphql-java version 17.4.
Locally I can build the 17.4 JAR file but my JAR file is only 1.6MB rather than 2.7 as in Maven Repo.

Cheers - Andy

**Comment by dondonz** [comment]
Hi @schaefa, I had a look at v17.4 and `sun.misc` is added to the `MAINFEST.MF` by the `biz.aQute.bnd.builder` plugin. graphql-java doesn't directly use anything from `sun.misc` so this may be imported indirectly.

I am not familiar with Sling or Feature Models. Does it look at the `MANIFEST.MF` file rather than the pom? If you have a longer error message or stack trace, could you please paste that here?

**Comment by schaefa** [comment]
I built a local file (./gradelw clean shadowJar) where I changed the line 155 in build.gradle to exclude sun.misc:

Import-Package: !com.google.*,!org.checkerframework.*,!javax.annotation.*,!graphql.com.google.*,!sun.misc,*

And that will result in a JAR file w/o that import. Unfortunately I am not able to build a maven artifact for 17.4 locally.
Can you provide me with instruction on how to do it (I tried './gradlew clean publishToMavenLocal  -x signGraphqlJavaPublication' but the JAR file's manifest does not contain the version number 17.4)?

When I used a JAR w/o sun.misc in it  Adobe AEM it runs just fine w/o any issues.

**Reply by dondonz** [reply]
I feel uneasy about publishing a local-only edited version, as you'll probably deploy this on other machines later, which won't have access to your edited version.

I inspected the MANIFEST.MF in both the latest release v20.0 and the latest master commit, both of these do not include `sun.misc`. Is it possible to update the version of graphql-java being used?

**Reply by schaefa** [reply]
This is the Import-Packages section of the MANIFEST.MF from the graphql-java-20.0.jar from 'https://repo1.maven.org/maven2/com/graphql-java/graphql-java/20.0/':

Import-Package: graphql,graphql.analysis,graphql.cachecontrol,graphql.
 collect,graphql.execution,graphql.execution.directives,graphql.execut
 ion.instrumentation,graphql.execution.instrumentation.dataloader,grap
 hql.execution.instrumentation.parameters,graphql.execution.preparsed,
 graphql.execution.reactive,graphql.i18n,graphql.introspection,graphql
 .language,graphql.normalized,graphql.parser,graphql.parser.antlr,grap
 hql.parser.exceptions,graphql.scalar,graphql.schema,graphql.schema.di
 ff,graphql.schema.diff.reporting,graphql.schema.diffing.ana,graphql.s
 chema.diffing.dot,graphql.schema.fetching,graphql.schema.idl,graphql.
 schema.idl.errors,graphql.schema.impl,graphql.schema.validation,graph
 ql.schema.visibility,graphql.util,graphql.validation,graphql.validati
 on.rules,org.dataloader;version="[3.2,4)",org.dataloader.stats;versio
 n="[3.2,4)",org.reactivestreams;version="[1.0,2)",org.slf4j;version="
 [1.7,2)",sun.misc
 
I cannot use a local built JAR file in the official Sling build but I used it here to see what works or not.

**Reply by dondonz** [reply]
Yes, you are right, it's there in the published JAR file.

My bad. I was generating the JAR with the Gradle build task on the v20.0 release commit and latest master, which strangely doesn't append `sun.misc`. I'll continue having a look.

**Reply by dondonz** [reply]
I should have run the shadowJar task instead.

**Reply by dondonz** [reply]
Thanks again for reporting the issue. I put up this PR to exclude `sun.misc.*` from the `MANIFEST.MF`. https://github.com/graphql-java/graphql-java/pull/3091

I'll separately have a look at the issue where MANIFEST.MF was not the first file listed

**Reply by schaefa** [reply]
It looks like that whatever is building the JAR file in shadowJar is not aware of that. When I extract the JAR file and then use 'jar -cvf' it will create the correct / expected JAR file with the MANIFEST.MF as first entry. 
Keep in mind that for example java.util.jar.JarInputStream is not finding the MANIFEST.MF if it is not the first entry (new JarInputStream(ins).getManifest() will return null).

**Reply by schaefa** [reply]
After some back and forth (I am not a gradle expect) I came up with this as a soluition (starting at line 143):

task extractWithoutGuava(type: Copy) {
    from({ zipTree({ "build/libs/graphql-java-${project.version}.jar" }) }) {
        exclude('/com/**')
    }
    into layout.buildDirectory.dir("extract")
}

task buildNewJar(type: Jar) {
    from layout.buildDirectory.dir("extract")
    archiveFileName = "graphql-java-tmp.jar"
    destinationDirectory = file("${project.buildDir}/libs")
    manifest {
        from file("build/extract/META-INF/MANIFEST.MF")
    }
    doLast {
        delete("build/libs/graphql-java-${project.version}.jar")
        file("build/libs/graphql-java-tmp.jar").renameTo(file("build/libs/graphql-java-${project.version}.jar"))
    }
}

shadowJar.finalizedBy extractWithoutGuava, buildNewJar

**Reply by schaefa** [reply]
I created a PR for my 'fix': https://github.com/graphql-java/graphql-java/pull/3097

**Reply by dondonz** [reply]
Thanks so much for the PR, I've merged it!

**Comment by schaefa** [comment]
I encounter another issue with release 17.4 and 20.0. All of the graphql-java JAR files cannot be used anymore as OSGi Bundles as the META-INF/MANIFEST.MF is not the first entry. This is caused by the shadowJar task. If I use task jar the order is correct.
If you use 'jar -tvf <jar file>' it will list all files and the MANIFEST.MF must be the first file to be listed. Right now it starts with Graphql.g4 and others.
Strongly enough I can import the bundle just fine through the OSGi console but not when deployed through the file system deployer.
